lib/fibered_mysql2/fibered_database_connection_pool.rb

```diff
@@ -0,0 +1,221 @@
+
+module FiberedMysql2
+  class FiberedConditionVariable < MonitorMixin::ConditionVariable
+    def initialize(monitor)
+      @monitor = monitor
+      @cond = EM::Synchrony::Thread::ConditionVariable.new
+    end
+  end
+
+  # From Ruby's MonitorMixin, with all occurrences of Thread changed to Fiber
+  module FiberedMonitorMixin
+    def self.extend_object(obj)
+      super
+      obj.__send__(:mon_initialize)
+    end
+
+    #
+    # Attempts to enter exclusive section.  Returns +false+ if lock fails.
+    #
+    def mon_try_enter
+      if @mon_owner != Fiber.current
+        @mon_mutex.try_lock or return false
+        @mon_owner = Fiber.current
+        @mon_count = 0
+      end
+      @mon_count += 1
+      true
+    end
+
+    #
+    # Enters exclusive section.
+    #
+    def mon_enter
+      if @mon_owner != Fiber.current
+        @mon_mutex.lock
+        @mon_owner = Fiber.current
+        @mon_count = 0
+      end
+      @mon_count += 1
+    end
+
+    #
+    # Leaves exclusive section.
+    #
+    def mon_exit
+      mon_check_owner
+      @mon_count -= 1
+      if @mon_count == 0
+        @mon_owner = nil
+        @mon_mutex.unlock
+      end
+    end
+
+    #
+    # Enters exclusive section and executes the block.  Leaves the exclusive
+    # section automatically when the block exits.  See example under
+    # +MonitorMixin+.
+    #
+    def mon_synchronize
+      mon_enter
+      begin
+        yield
+      ensure
+        begin
+          mon_exit
+        rescue => ex
+          ActiveRecord::Base.logger.error("Exception occurred while executing mon_exit: #{ex}")
+        end
+      end
+    end
+    alias synchronize mon_synchronize
+
+    #
+    # Creates a new FiberedConditionVariable associated with the
+    # receiver.
+    #
+    def new_cond
+      FiberedConditionVariable.new(self)
+    end
+
+    private
+
+    # Initializes the FiberedMonitorMixin after being included in a class
+    def mon_initialize
+      @mon_owner = nil
+      @mon_count = 0
+      @mon_mutex = EM::Synchrony::Thread::Mutex.new
+    end
+
+    def mon_check_owner
+      @mon_owner == Fiber.current or raise FiberError, "current fiber not owner"
+    end
+
+    def mon_enter_for_cond(count)
+      @mon_owner = Fiber.current
+      @mon_count = count
+    end
+
+    # returns the old mon_count
+    def mon_exit_for_cond
+      count = @mon_count
+      @mon_owner = nil
+      @mon_count = 0
+      count
+    end
+  end
+
+  module FiberedDatabaseConnectionPool
+    include FiberedMonitorMixin
+
+    module Adapter_4_2
+      def cached_connections
+        @reserved_connections
+      end
+    end
+
+    module Adapter_5_2
+      def cached_connections
+        @thread_cached_conns
+      end
+    end
+
+    case Rails::VERSION::MAJOR
+    when 4
+      include Adapter_4_2
+    when 5, 6
+      include Adapter_5_2
+    end
+
+    def initialize(connection_spec)
```

Comment:
I remember you mentioning that this was having some issues because it was loosing track of arguments that were being passed by other layers that were mixed in on top of this gem.  We can make this a little more lenient in that regard by doing

```ruby
def initialized(connection_spec, *_args, **_keyword_args)
  # do the connection_spec config check here
  super # <- by doing this without arguments it passing everything that was passed to this method up the chain
  # do the rest of the method's job
end
```

Reply:
@jebentier I'm having trouble implementing this suggestion. I'm getting errors like this:
```
     Failure/Error: super

     ArgumentError:
       wrong number of arguments (given 2, expected 1)
     # ./lib/fibered_mysql2/fibered_database_connection_pool.rb:146:in `initialize'
     # ./spec/unit/fibered_database_connection_pool_spec.rb:357:in `block (3 levels) in <top (required)>'
     # ./spec/unit/fibered_database_connection_pool_spec.rb:374:in `block (3 levels) in <top (required)>'
```

It looks like `_args` evaluates to `[]` and `_keyword_args` evaluates to `{}`, and one of those is being passed to super. I put this hack in there to account, but does that behavior seem right?
```
      if _args.empty? && _keyword_args.empty?
        super(connection_spec)
      else
        super
      end
```

Reply:
Ah, I guess `super` without passing args isn't smart enough to re-splat the keyword args ports. Instead of the hack you put in, you can do the following:

`super(connection_spec, *args, **keyword_args)`